### PR TITLE
Fixed ordering bug

### PR DESCRIPTION
**Pull request by stephenwf**
@giorgiosironi @thewilkybarkid 

This should fix the ordering of the recommendations results. I've been stepping through how the rules resolve back out and I can't find any inconsistencies at the moment. Can you provide any URLs from the prod/end2end where the data is incorrect?

Thanks

### CODE REVIEW

**Comment by thewilkybarkid** [comment]
This changes it so all recommendations are ordered by date? It should be by rules, then date (say if the article appears in two collections, the most recent collection would appear first, but if it's related to an article published before either of the collections, that would still appear first).

**Comment by stephenwf** [comment]
In that case the question becomes: https://github.com/elifesciences/recommendations/blob/develop/src/App/Kernel.php#L267-L295

Is this the correct order

**Comment by giorgiosironi** [comment]
I'll look for a production link, since in end2end I was seeing articles not being related to each other.

**Comment by giorgiosironi** [comment]
```
curl -v https://prod--gateway.elifesciences.org/articles/15893/related | jq "."
```
gives me
```
[
  {
    "id": "11911",
       ...
```
as the only result.
But
```
curl -v https://prod--gateway.elifesciences.org/recommendations/article/15893 | jq ".items[].id"
```
gives
```
"25159"
"08231"
```
which is not the related article. We expect to see 11911 there, probably at the top.

**Comment by stephenwf** [comment]
This was intentional, according to the original spec we do not show related `short-reports`, unless they happen to be the most recent article: https://github.com/elifesciences/recommendations-formula/blob/master/META/TASKS.md

**Comment by thewilkybarkid** [comment]
Ah, 15893 is a correction of a short report. I think we should insert a new rule at 11 (ie after all the other article relation rules) that adds any related articles not already included to catch these.

(We're not actually displaying recommendations for correction articles anyhow, but there might be other legitimate cases.)

**Comment by thewilkybarkid** [comment]
Though actually that short report should have triggered rule 3.

**Comment by giorgiosironi** [comment]
I'm using two copies of 15893 related to each other because they are very small articles without figures, so publishing both of them in tests takes a shorter amount of time. However if there is a pair of better articles to use for tests in end2end in here:
https://github.com/elifesciences/elife-spectrum/tree/master/spectrum/templates
then tell me the two ids and I'll use them.

**Comment by stephenwf** [comment]
So I'm trying to work out how this should work. The related articles are two way, so each appears in each others. So if the rule: 

> Articles an Insight, Feature, Editorial, Research Advance, Research Exchange or Registered Report, Retraction or Correction is linked to

was meant to catch them, then it would add a relation to all items regardless of their type.

Currently we are treating the relations as bi-directional so they appear in each others recommendations, much like the related items. Is this incorrect? Should we be adding both sides of the relationship separately with their own rules?

**Comment by thewilkybarkid** [comment]
Hmm, 3 is the inverse of 1–2/4–10. But, I think we can lose it, and as long as there's a rule after these that catches articles that hasn't yet been included.

(We're currently not including the _type_ of relationship in the related articles endpoint (eg 'corrected by', 'commentary on'), I think these rules might have been a bit clearer with them.)

**Comment by stephenwf** [comment]
That will get interesting too, opens questions of overlapping. I was going to, for debugging purposes, add a field like that in

**Comment by stephenwf** [comment]
Okay so in order to catch related articles, we will need to make a call to `/articles/{id}/related` at runtime. If we want to cache this we will need know when to invalidate and implement something to cover that. Currently we have `GET /articles/{id}` cacheable and invalidated by SQS `{"type": "article", "id": "{id}"}`

**Comment by thewilkybarkid** [comment]
You can use the same event.

**Comment by stephenwf** [comment]
@thewilkybarkid @giorgiosironi I've changed the rules here now so they will work as expected. I also ran into the database bug so fixed that too!

File: src/App/Console.php [hunk 1]
```diff
@@ -59,6 +72,42 @@ public function __construct(Application $console, Kernel $app)
...
 
+    public function queuePushCommand(InputInterface $input, OutputInterface $output)
+    {
+        $id = $input->getArgument('id');
+        $type = $input->getArgument('type');
+        // Enqueue.
+        $this->enqueue($type, $id);
+    }
+
+    private function enqueue($type, $id)
```
Comment (giorgiosironi) [inline src/App/Console.php]: one day these commands should go into bus-sdk-php, but not now

**Comment by giorgiosironi** [comment]
I'll merge the puli one first and merge develop into here, to make sure it's not Puli failing the build.

**Comment by giorgiosironi** [comment]
It's failing on database creation/deletion/regeneration now:
```
10:50:33 [54.174.249.239] out:           ID: recommendations-create-database
10:50:33 [54.174.249.239] out:     Function: cmd.run
10:50:33 [54.174.249.239] out:         Name: ./bin/console generate:database --drop --env=ci
10:50:33 [54.174.249.239] out:       Result: False
10:50:33 [54.174.249.239] out:      Comment: Command "./bin/console generate:database --drop --env=ci" run
10:50:33 [54.174.249.239] out:      Started: 10:50:27.041212
10:50:33 [54.174.249.239] out:     Duration: 127.461 ms
10:50:33 [54.174.249.239] out:      Changes:   
10:50:33 [54.174.249.239] out:               ----------
10:50:33 [54.174.249.239] out:               pid:
10:50:33 [54.174.249.239] out:                   23478
10:50:33 [54.174.249.239] out:               retcode:
10:50:33 [54.174.249.239] out:                   1
10:50:33 [54.174.249.239] out:               stderr:
10:50:33 [54.174.249.239] out:                   
10:50:33 [54.174.249.239] out:                                                                                                  
10:50:33 [54.174.249.239] out:                     [Doctrine\DBAL\Exception\TableExistsException]                               
10:50:33 [54.174.249.239] out:                     An exception occurred while executing 'CREATE TABLE Rules (rule_id CHAR(36)  
10:50:33 [54.174.249.239] out:                      NOT NULL COMMENT '(DC2Type:guid)', id VARCHAR(64) NOT NULL, type VARCHAR(6  
10:50:33 [54.174.249.239] out:                     4) NOT NULL, published DATETIME DEFAULT NULL, isSynthetic TINYINT(1) DEFAUL  
10:50:33 [54.174.249.239] out:                     T '0' NOT NULL, PRIMARY KEY(rule_id)) DEFAULT CHARACTER SET utf8 COLLATE ut  
10:50:33 [54.174.249.239] out:                     f8_unicode_ci ENGINE = InnoDB':                                              
10:50:33 [54.174.249.239] out:                     SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'Rules' alre  
10:50:33 [54.174.249.239] out:                     ady exists                                                                   
10:50:33 [54.174.249.239] out:                                                                                                  
10:50:33 [54.174.249.239] out:                   
10:50:33 [54.174.249.239] out:                                                                                                  
10:50:33 [54.174.249.239] out:                     [Doctrine\DBAL\Driver\PDOException]                                          
10:50:33 [54.174.249.239] out:                     SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'Rules' alre  
10:50:33 [54.174.249.239] out:                     ady exists                                                                   
10:50:33 [54.174.249.239] out:                                                                                                  
10:50:33 [54.174.249.239] out:                   
10:50:33 [54.174.249.239] out:                                                                                                  
10:50:33 [54.174.249.239] out:                     [PDOException]                                                               
10:50:33 [54.174.249.239] out:                     SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'Rules' alre  
10:50:33 [54.174.249.239] out:                     ady exists                                                                   
10:50:33 [54.174.249.239] out:                                                                                                  
10:50:33 [54.174.249.239] out:                   
10:50:33 [54.174.249.239] out:                   generate:database [-d|--drop]
10:50:33 [54.174.249.239] out:               stdout:
10:50:33 [54.174.249.239] out:                   [2017-03-16 10:50:27] recommendations-api.ERROR: An exception occurred while executing 'CREATE TABLE Rules (rule_id CHAR(36) NOT NULL COMMENT '(DC2Type:guid)', id VARCHAR(64) NOT NULL, type VARCHAR(64) NOT NULL, published DATETIME DEFAULT NULL, isSynthetic TINYINT(1) DEFAULT '0' NOT NULL, PRIMARY KEY(rule_id)) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB':  SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'Rules' already exists {"exception":"[object] (Doctrine\\DBAL\\Exception\\TableExistsException(code: 0): An exception occurred while executing 'CREATE TABLE Rules (rule_id CHAR(36) NOT NULL COMMENT '(DC2Type:guid)', id VARCHAR(64) NOT NULL, type VARCHAR(64) NOT NULL, published DATETIME DEFAULT NULL, isSynthetic TINYINT(1) DEFAULT '0' NOT NULL, PRIMARY KEY(rule_id)) DEFAULT CHARACTER SET utf8 COLLATE utf8_unicode_ci ENGINE = InnoDB':\n\nSQLSTATE[42S01]: Base table or view already exists: 1050 Table 'Rules' already exists at /srv/recommendations/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php:49, Doctrine\\DBAL\\Driver\\PDOException(code: 42S01): SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'Rules' already exists at /srv/recommendations/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:59, PDOException(code: 42S01): SQLSTATE[42S01]: Base table or view already exists: 1050 Table 'Rules' already exists at /srv/recommendations/vendor/doctrine/dbal/lib/Doctrine/DBAL/Driver/PDOConnection.php:57)"}
10:50:33 [54.174.249.239] out: ----------
```

**Comment by giorgiosironi** [comment]
Pleases to see this green, shall we merge?

**Comment by thewilkybarkid** [comment]
👍

**Comment by stephenwf** [comment]
Good to go 👍

**Comment by giorgiosironi** [comment]
We'll merge the bot new activity first, then this.

**Comment by thewilkybarkid** [comment]
Think we can merge now?

**Pull request merged**